UnitTests/GitCommandsTests/Submodules/SubmoduleStatusProviderTests.cs

```diff
@@ -0,0 +1,174 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using CommonTestUtils;
+using FluentAssertions;
+using GitCommands;
+using GitCommands.Submodules;
+using GitUIPluginInterfaces;
+using Microsoft.VisualStudio.Threading;
+using NUnit.Framework;
+
+namespace GitCommandsTests.Submodules
+{
+    [SetCulture("en-US")]
+    [SetUICulture("en-US")]
+    [TestFixture]
+    internal class SubmoduleStatusProviderTests
+    {
+        public class IntegrationTests
+        {
+            private GitModuleTestHelper _repo1;
+            private GitModuleTestHelper _repo2;
+            private GitModuleTestHelper _repo3;
+
+            // Note that _repo2Module and _repo3Module point to the submodules under _repo1Module,
+            // not _repo2.Module and _repo3.Module respectively. In general, the tests should here
+            // should interact with these modules, not with _repo2 and _repo3.
+            private GitModule _repo1Module;
+            private GitModule _repo2Module;
+            private GitModule _repo3Module;
+
+            [SetUp]
+            public void SetUp()
+            {
+                _repo1 = new GitModuleTestHelper("repo1");
+                _repo2 = new GitModuleTestHelper("repo2");
+                _repo3 = new GitModuleTestHelper("repo3");
+
+                _repo2.AddSubmodule(_repo3, "repo3");
+                _repo1.AddSubmodule(_repo2, "repo2");
+                var submodules = _repo1.GetSubmodulesRecursive();
+
+                _repo1Module = _repo1.Module;
+                _repo2Module = submodules.ElementAt(0);
+                _repo3Module = submodules.ElementAt(1);
+            }
+
+            [TearDown]
+            public void TearDown()
+            {
+                _repo1.Dispose();
+                _repo2.Dispose();
+                _repo3.Dispose();
+            }
+
+            [Test]
+            public void UpdateSubmoduleStatus_valid_result_for_top_module()
+            {
+                ISubmoduleStatusProvider provider = new SubmoduleStatusProvider();
```

Comment:
this line is repeated several times, you could move it into `Setup` (it is run before each test)